### PR TITLE
Migrate prehandle workflow to new states implementation

### PR DESCRIPTION
**Pull request by netopyr**
This change adapts the prehandle workflow to the new states design.
- It expects a `HederaState` as the root node of the state.
- It ensures that all access to state is tracked. For instance all stores that are passed to the handlers are being tracked.
- It attaches the read keys to the `TransactionMetadata`.

I changed the design after our discussion, which may be controversial. Please have a look.
- The handlers are not aware of `TransactionMetadata`, because it contains information, a service should not know about (the read keys).
- The handlers get a context object `PrehandleHandlerContext`. (I know, the name needs to be improved... 😁) This context contains all information passed to the handlers (`transactionBody` and `payer`). And it gathers information from the handlers (`responseCode`, `requiredNonPayerKeys`, and `handlerMetadata`).
- I removed `TransactionMetadataBuilder` as it was not needed anymore.

### CODE REVIEW

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (hendrikebbers) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:49]: Should this be a generic class to not work with Object?

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (hendrikebbers) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:48]: I do not like that but have no good idea about it

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (hendrikebbers) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:93]: Should the returned list be mutable?

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (hendrikebbers) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:62]: null check for context

File: hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java
Comment (hendrikebbers) [inline hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java:28]: Magic Number. Is it just a random number or is 42 any special account id?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java:51]: Please adopt the pattern that @rbair23 started in this PR: https://github.com/hashgraph/hedera-services/pull/4696/files#diff-d92d4de55a0434acb2c0e3a442dc4a896bfdd387e68df62945984269a218f926

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java:145]: null check for storeSupplier

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:126]: Is always null.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:167]: null check usedStates

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:173]: Maybe you can extract a method? The stream handling becomes hard to read.

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/dispatcher/DispatcherTest.java
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/dispatcher/DispatcherTest.java:49]: Please configure IntelliJ to not use start imports

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:49]: In that case, we would have to work with `TransactionMetadata<?>` pretty much everywhere, i.e. more complexity without any advantages.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:93]: No, good catch.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:62]: `context.getTxn()` is a null check... kind of. 😏 

Do you prefer `requireNonNull(context).getTxn()`?

File: hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java:28]: Just some random test data.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java:51]: Sure, once it is merged.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/ingest/IngestWorkflowImpl.java:145]: Added

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:126]: Right... deleted by mistake. Fixed

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:167]: It's a private method. I don't think that this is needed.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:173]: Done

File: hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/dispatcher/DispatcherTest.java
Comment (netopyr) [reply hedera-node/hedera-app/src/test/java/com/hedera/node/app/workflows/dispatcher/DispatcherTest.java:49]: Done

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:48]: I think `TransactionMetadata` record is clearer than Object

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:278]: `toString` still shows `TransactionMetadataBuilder`

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:49]: I think you can add `TransactionMetadata` here instead of object

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:101]: This name can be changed to differentiate the `HederaKeys` to the keys here which are account numbers. 
Probably `AccessedAccounts` or `ReadAccountNumbers` ?

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:92]: Need to add more javadoc to avoid confusion on `keys`, which mean account numbers

File: hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java:28]: I see `ErrorTransactionMetadata` is deleted. If so, can we delete this test ?

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:38]: There is one class named `PreHandleContext` and this class named `PreHandlerContext` . Should we merge both ?

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:49]: That would mean, that all handlers can only attach a `TransactionMetadata` object. This seems too limited, in particular, if we want to prepare more work during pre-handle in the future.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java:210]: Should we probably move these methods to a different class , since we will get more stores ?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:165]: Do we need all the below methods to be static ? Please add javadocs to the below methods

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:156]: We should probably use `createInvalidTransactionMetadata` or so, to be consistent with `createTransactionMetadata`

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:101]: Not sure, if I understand you correctly. `readKeys` are not just account numbers. They can point to basically any value that is stored anywhere in the Merkle tree.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/ReadableStatesTracker.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/ReadableStatesTracker.java:25]: Please add javadocs. Is this a factory similar to `StoreCache`  we had previously

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:48]: Currently we don't have any other transactions other than schedules. Object is not clear.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:101]: Yes. It can be any entity, but they will not be `HederaKeys`. `ReadKeys` will be confusing to know they are entity numbers and not cryptographic keys imo.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:48]: But we have discussed that in one of our meetings. We agreed to use `Object` or a marker interface to be flexible.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:278]: Good catch. Fixed.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:341]: Fixed

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:101]: We use `readKeys` in the states API. Maybe we should discuss it in the larger group. I am ok with changing it, but it needs to be consistent with the states API.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java:92]: Right, let's put it on the list of topics to discuss tomorrow. Once we have agreed on the wording, documenting it will be the easy part. 😄

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:38]: I did not create `PrehandleContext` and therefore I was not sure, what the intention was for that class. Maybe it has become obsolete with the new design and we can happily repurpose it.

File: hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/test/java/com/hedera/node/app/spi/test/meta/ErrorTransactionMetadataTest.java:28]: I think even though the class was changed, the functionality is still there. I will take a look and merge the different test classes into one.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/dispatcher/TransactionDispatcher.java:210]: Good idea

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:38]: I think we need the account numbers and file numbers in some prehandle methods. So I created that class. I think we can merge both. Need not be done in this PR a separate PR

**Review by tinker-michaelj — APPROVED** [review]
LGTM

**Review by Neeharika-Sompalli — APPROVED** [review]
Looks great. All the comments can be added in a new PR after discussion needed. Need not block this PR.

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java
Comment (netopyr) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/PrehandleHandlerContext.java:38]: Will do

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:165]: Will do

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:156]: Ok, will change it.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/ReadableStatesTracker.java
Comment (netopyr) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/ReadableStatesTracker.java:25]: Will do

**Pull request merged**
